libkatana_python_native/include/katana/python/NumbaSupport.h

```diff
@@ -14,138 +14,207 @@ namespace detail {
-  //  or name template argument of some kind.
-
-  // Overload to handle non-const methods on classes.
-  template <typename Cls, typename Return, typename ImplCls, typename... Extra>
-  void operator()(
-      pybind11::class_<Cls>& cls, const char* name,
-      Return (ImplCls::*pmf)(Args...), const Extra&... extra) const {
-    // This leaks a single pointer sized struct for each defined numba function.
-    // Repeated import could theoretically cause this to matter, but it's very
-    // unlikely.
-    def_class_method<Cls, Return>(
-        cls, name, pmf, new MemberFunction<ImplCls, Return, Args...>{pmf},
-        extra...);
-  }
+template <typename Cls, typename Return, typename... Args>
+constexpr auto
+MemberFunctionBinderInferer(Return (Cls::*)(Args...) const) {
+  return ConstMemberFunctionBinderImpl<Cls, Return, Args...>{};
+}
 
-  // Overload to handle const methods on classes.
-  template <typename Cls, typename Return, typename ImplCls, typename... Extra>
-  void operator()(
-      pybind11::class_<Cls>& cls, const char* name,
-      Return (ImplCls::*pmf)(Args...) const, std::true_type,
-      const Extra&... extra) const {
-    // As above.
-    def_class_method<Cls, Return>(
-        cls, name, pmf, new ConstMemberFunction<ImplCls, Return, Args...>{pmf},
-        extra...);
-  }
-};
+template <auto member_func>
+using MemberFunctionBinder = decltype(MemberFunctionBinderInferer(member_func));
 
 }  // namespace detail
 
-/// Declare a method or function to be called from Numba and Python.
+/// Define a Python and numba method.
+/// This is used as follows:
+/// @code
+/// DefWithNumba&lt;&func>(cls, "func");
+/// @endcode
+/// If the method is overloaded, you must select a specific overload using
+/// pybind11::overload_cast().
+/// @code
+/// DefWithNumba&lt;pybind11:overload_cast&lt;const std::string&>(&func)>(cls, "func");
+/// @endcode
 ///
-/// This should be called the same way `pybind11`'s `def` function is called.
+/// Numba interfacing does not (yet) support exposing overloading itself to
+/// Numba code.
 ///
-/// \tparam Args The argument types of the function.
-template <typename... Args>
-constexpr detail::DefWithNumbaImpl<Args...> DefWithNumba{};
+/// \tparam func The function pointer @b value (not type).
+/// \param cls The python class object.
+/// \param name The name of the method in python.
+/// \param extra Extra arguments as you would pass to pybind11::class_::def().
+template <auto func, typename... Extra>
+constexpr void
+DefWithNumba(pybind11::module_& m, const char* name, const Extra&... extra) {
+  static_assert(
+      std::is_same_v<
+          typename detail::StaticFunctionBinder<func>::StaticFunctionPointer,
+          decltype(func)>);
+  detail::StaticFunctionBinder<func>::template def_method<func>(
+      m, name, extra...);
+}
+
+/// Define a Python and numba method.
+/// This is used as follows:
+/// @code
+/// DefWithNumba&lt;&Cls::method>(cls, "method");
+/// @endcode
```

Comment:
nit: Do you have to say `&lt` to represent `<` ?

Reply:
Yes. Yes. I do. Doxygen is a PITA.